### PR TITLE
[FIX] tools, *: ignore the active field when updating assets

### PR DESCRIPTION
**Pull request by bso-odoo**
Before this commit when a module was updated all ir.asset records were
reset to their defined `active` state, if defined.
This causes assets related to old snippet versions to be made inactive
even if those old snippet versions are used inside existing pages.

It used to work when the activation of assets was made through view
inheritance because when views are defined through a `<template>` tag,
the `active` attribute is in fact ignored during updates since [1],
except for new records since [2].

This commit introduces an `<asset>` tag in the XML import format.
It is an alias of `<record ... model="ir.asset">` with the additional
feature that it avoids taking the `active` field into account during
updates for existing `ir_asset` records, just like `<template>` if the
`active` field is mentioned as attribute of the tag.
We then rely on the `website_disable_unused_snippets_assets` cron to
properly disable any unused asset at a later stage (note that the bug
being fixed here was mitigated by the fact that cron also re-enabled
assets which were disabled by mistake... but that might happen only a
few days later).

Another approach was to overload `_load_records_write` in `base`'s
`ir_asset.py` to avoid taking the `active` field into account when
updating records:
```py
    def _load_records_write(self, values):
        values.pop('active', None)
        super()._load_records_write(values)
```
But this is not as stable because it changes the way `ir.asset` records
are imported when the `<record>` tag is used. In the end we chose to be
consistent and do exactly the same as `<template>`, as this also allows
more and should be entirely stable.

[1]: https://github.com/odoo/odoo/commit/2d296cb77922d33be2dc45b900191fac34bda429#diff-175c28787c272a219b9275f79262a48af9aa029e718f45077fd609737559e84eR803-R804
[2]: https://github.com/odoo/odoo/commit/f1c70d4cc943ac4eb81a85a9dc005de34cd2060a#diff-175c28787c272a219b9275f79262a48af9aa029e718f45077fd609737559e84eR801-R804

task-2963840

(Follow-up of https://github.com/odoo/upgrade/pull/3829)

### CODE REVIEW

**Review by rdeodoo — COMMENTED** [review]
Can't you use `_load_records_write()` instead and pop the `active` field all the time?

**Comment by bso-odoo** [comment]
> Can't you use `_load_records_write()` instead and pop the `active` field all the time?

TL;DR: I simplified as suggested.

@rdeodoo As far as I understood: `_load_records_write()` is called with 2 kinds of records:
- those with a defined `id`:
```py
                if vals.get('id'):
                    data['record'] = self.browse(vals['id'])
                    to_update.append(data)
```
- those, within an update, that have an existing `xml_id` and are allowed to be updated:
```py
                if not (update and d_noupdate):
                    to_update.append(data)
```
I have not seen imports with explicit `id`s yet, but maybe those could be used to toggle the `active` state of an asset.
On the other hand, we know we want to rely on the cron to disable assets anyway, so I guess we can always skip the `active` during import writes.

File: odoo/addons/base/models/ir_asset.py [hunk 1]
```diff
@@ -392,6 +392,10 @@ def _process_command(self, command):
             target = None
         return directive, target, path_def
 
+    def _load_records_write(self, values):
+        values.pop('active', None)
+        super()._load_records_write(values)
```
Comment (qsm-odoo) [inline odoo/addons/base/models/ir_asset.py:397]: @odoo/rd-security @odoo/rd-framework-py What do you think of this? The discussion started here: https://github.com/odoo/upgrade/pull/3829#issuecomment-1297363594 As suggested, maybe we should also change the way this is done for views (not relying on the use of `<template>`, but doing it like here) ? https://github.com/odoo/odoo/blob/f1c70d4cc943ac4eb81a85a9dc005de34cd2060a/openerp/tools/convert.py#L801

@bso-odoo Could be nice to mention more things in the commit message (the fact that it was the case for "assets views", the link to the code doing that, the link of the original commit I mentioned which added the active fields, etc ;))

**Comment by bso-odoo** [comment]
@qsm-odoo Improved commit message.

File: odoo/addons/base/models/ir_asset.py (hunk 1 above)
Comment (xmo-odoo) [reply odoo/addons/base/models/ir_asset.py:397]: > What do you think of this?

:shrug: it's not great but I guess the updating the assets via the XML files is not really a use-case?

> As suggested, maybe we should also change the way this is done for views (not relying on the use of `<template>`, but doing it like here) ?

Is there a use-case for that? Because it seems a lot less innocuous, as I could well see the ability to update view activation being important for module evolution / migration, and if a user wants to craft their updates around this, they can use custom import `context` and `<record>` tags.

Comment (xmo-odoo) [reply odoo/addons/base/models/ir_asset.py:397]: > Maybe not (although could make sense as explained above), but in that case, should we rely on the use of `<template>`?

We've been doing that for 8 years so... maybe?

>  I don't see a good reason to not have the same behavior if that qweb view is defined using `<record>`?

And I don't see any reason to have it. If you're using raw `record` you can already achieve that using `default_` keys in the context.

Comment (qsm-odoo) [reply odoo/addons/base/models/ir_asset.py:397]: > We've been doing that for 8 years so... maybe?

And hopefully we improved the codebase for 8 years :wink: 

> If you're using raw record you can already achieve that using default_ keys in the context.

I don't know what you mean, can you explain? :grimacing: 

> And I don't see any reason to have it.

Probably not critical but I have 3 :smile: 
- Consistency: I really don't understand why would `<template>` be different from `<record>`. I am pretty convinced that most users only see it as a shortcut for `<record>`, not something with additional behaviors. I would be surprised if the error was not there in custom modules. And as we plan to develop the documentation about creating custom website modules/themes to attract developers and improve our PS tech designer team tools, the more consistent the better.
- Readability: I don't know how much time he took me to find that piece of code but way too long (and actually, I found it 2 years ago and it took me about the same time to find it again last week).
- Robustness: Kinda related to the previous point but having this behavior anchored to the views python file via a dedicated python method would prevent this behavior to be lost / not adapted in a refactoring. Maybe if it was there, the behavior would not have been lost when creating the `ir.asset` model.

Again nothing critical (I would actually not "fix" that in stable, but in master).

Comment (xmo-odoo) [reply odoo/addons/base/models/ir_asset.py:397]: > > If you're using raw record you can already achieve that using default_ keys in the context.
> 
> I don't know what you mean, can you explain? grimacing

`default_` keys in contexts only apply to create calls. The record (and function) elements support `uid` and `context` attributes to customise their environment.

> * Consistency: I really don't understand why would `<template>` be different from `<record>`. I am pretty convinced that most users only see it as a shortcut for `<record>`

Which it is.

> not something with additional behaviors.

The two are not exclusive? `<template>` is strictly less flexible than `<record>`. That means if there are specific use cases were want to support in `<template>` they may need to... be specifically supported there. That doesn't mean `<record>` needs them.

Meanwhile if you move this to the model you're now fundamentally altering the way it works globally.

Comment (qsm-odoo) [reply odoo/addons/base/models/ir_asset.py:397]: I hear you, but there is no way lambda developers guess that and even with a good tutorial, I still don't think it's good in practice: having to explain that you can use `<record>` to create records but for qweb views, there is high probability that you should use `<template>` to not break future updates and that even in the case you would want to use `<record>` for your specific usecase, it would not work as you would have to write an upgrade script to make it work... meh.

> Meanwhile if you move this to the model you're now fundamentally altering the way it works globally.

**during updates** (which is my point: updating the active field of a qweb view during updates does not work ever for frontend related layouts without upgrade scripts, so I would always disabling it by default during update). But ok, I guess we should not support qweb views are only used for frontend layouts.

cc @bso-odoo If you can at least advertise that behavior next to the `active` field definition of the ir.ui.view model ("is not updated during updates if `<template>` is used instead of `<record>`, see `_tag_template`" something like that), that would be nice :+1:). (An other PR in 14.0 mentioning that discussion).

Comment (qsm-odoo) [reply odoo/addons/base/models/ir_asset.py:397]: @rdeodoo For this current PR here, with `ir.asset`, what do you think? Should we use `_load_records_write` as in the current state, or should we try to use `default_` keys in context definition of all `<record>` to indicate `active` should only be set to `False` during creation and not update. In the end, if it makes sense to keep the standard behavior of `<record>` for views, I think it makes sense to keep it for `ir.asset` too. Or we could also create `<asset>`, to behave just like `<template>`. Probably what I would vote for: that way it's consistent.

Comment (rdeodoo) [reply odoo/addons/base/models/ir_asset.py:397]: In order of preference:
1. Have `<template/>` and `<record/>` behaves the same (the active field bypass also occurs with `<record/>`), it's very confusing that the shortcut version is by passing this field, it is just bug prone.
And use `_load_records_write` for both I guess or something like that
2. If framework is against is and wants to keep the inconsistency between `<record/>` and `<template/>` , indeed an `<asset>` tag to mimick that `active` field bypass with the shortcut tag would be the second best choice to keep to consistency
3. current state with only `_load_records_write`
4. `default_` stuff

Comment (rdeodoo) [reply odoo/addons/base/models/ir_asset.py:397]: @bso-odoo You can adapt with second choice.
If the framework accepts to align `<template/>` and `<record/>` later we will easily be able to do the same, + it's always nice to have a shortcut for this

Comment (bso-odoo) [reply odoo/addons/base/models/ir_asset.py]: @rdeodoo I introduced the `<asset>` tag, and adapted all (~100) existing `<record ... model="ir.asset">` records.
I will have to make sure I scan for those again at each forward-port.
Note that there was one such record in enterprise => odoo/enterprise#35153.

The rule is the same as for the `<template>` tag:
- if `active` is defined as an attribute on the `<asset>` node, apply it only for new records (otherwise, do not talk about it in the created record)

Comment (qsm-odoo) [reply odoo/addons/base/models/ir_asset.py]: @bso-odoo Don't you need to do it for themes too alongside making sure it works for theme.ir.asset as well? :wink: Or is this working in themes already because they are theme.ir.asset? (I guess the current theme.ir.asset are compatibility stuff so not as important (although still need to be fixed the same way if the bug is the same) but custom themes would create snippets using that)

File: odoo/import_xml.rng [hunk 2]
```diff
@@ -214,6 +214,24 @@
         </rng:element>
     </rng:define>
 
+    <rng:define name="asset">
+        <rng:element name="asset">
+            <rng:attribute name="id" />
+            <rng:attribute name="name" />
+            <rng:optional><rng:attribute name="active" /></rng:optional>
+            <rng:optional><rng:attribute name="website_id" /></rng:optional>
```
Comment (rdeodoo) [inline odoo/import_xml.rng:222]: Do we need this one? It's probably something we won't be using, like for views, I am surprised it exists, probably some old stuff not used anymore
I'd leave that to the normal `<field/>` behavior, no need to add it as attribute capability IMO

File: odoo/addons/base/models/ir_asset.py (hunk 1 above)
Comment (rdeodoo) [reply odoo/addons/base/models/ir_asset.py]: Themes indeed need to be adapted and we should also convert those to `<assets/>` and adapt the `_tag_asset` new method to handle it like the `_tag_template` does for templates -> Map `<assets/>` to `theme.ir.asset` if declared in a theme module
```py
        if self.module.startswith('theme_'):
            model = 'theme.ir.asset'
        else:
            model = 'ir.asset'
```

Comment (bso-odoo) [reply odoo/addons/base/models/ir_asset.py]: @qsm-odoo @rdeodoo Now handling `theme.ir.asset` too, additional PR: https://github.com/odoo/design-themes/pull/629.
Also note that I added `directive` as an attribute of `bundle`.

**Review by rdeodoo — APPROVED** [review]
The approach is LGTM.
Didn't test nor properly reviewed, I'll let @qsm-odoo decide if he wants to review/merge it himself before doing so.

Since this will be introduced in stable, I'd just add a test somehow, at least for the regular website part (<assets> tag being ir.asset), ideally also for themes (<assets> tag being theme.ir.asset).
Maybe worth looking into: `test_qweb_cdn`'s `_load()`.

**Comment by qsm-odoo** [comment]
I'll check once the test addition will have been looked into :wink: 

cc @rdeodoo @bso-odoo

**Comment by bso-odoo** [comment]
> I'd just add a test somehow, at least for the regular website part ( tag being ir.asset), ideally also for themes ( tag being theme.ir.asset).

@rdeodoo @qsm-odoo I have added the tests: `base/tests/test_ir_asset.py` for `ir.asset` and `website/tests/test_theme_ir_asset.py` for `theme.ir.asset`...
But it's almost a copy/paste of each other. If you prefer, I can put both inside website and factorize a bit.

File: addons/website/tests/asset_tag.xml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <asset id="theme_default.test_asset_tag_aaa" name="Test asset tag (init active => keep active => update active)">
```
Comment (rdeodoo) [inline addons/website/tests/asset_tag.xml:4]: we don't create tests/fake data in real modules, this should be move to `test_website` which only purpose is exactly to do this (see manifest)

File: odoo/addons/base/tests/asset_tag.xml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <asset id="base.test_asset_tag_aaa" name="Test asset tag (init active => keep active => update active)">
+        <bundle>test_asset_bundle</bundle>
+        <path>base/tests/something.scss</path>
+    </asset>
+    <asset id="base.test_asset_tag_aii" name="Test asset tag (init active => make inactive => update inactive)">
```
Comment (rdeodoo) [inline odoo/addons/base/tests/asset_tag.xml:8]: Triple check how are fake/tests data done in base module, probably to be moved too, didn't check

File: odoo/addons/base/tests/test_ir_asset.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+class TestAsset(TransactionCase):
+
+    def test_00_asset_tag(self):
+        """
+        Verify that assets defined with the <asset> tag are properly imported.
+        """
+        # Load new records
+        convert_file(self.cr, 'base',
+            get_module_resource('base', 'tests', 'asset_tag.xml'),
+            {}, 'init', False, 'test')
```
Comment (rdeodoo) [inline odoo/addons/base/tests/test_ir_asset.py:18]: image
```suggestion
        convert_file(
            self.cr, 'base',
            get_module_resource('base', 'tests', 'asset_tag.xml'),
            {}, 'init', False, 'test'
        )
```

File: odoo/addons/base/tests/test_ir_asset.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        prepend_asset.bundle = 'changed'
+        prepend_asset.path = 'base/tests/changed.scss'
+        asset_with_extra_field.sequence = 3
+        active_switch_asset_reset.active = False
+        active_switch_asset_ignore.active = False
+        inactive_switch_asset.active = True
+
+        # Update records
+        convert_file(self.cr, 'base',
+            get_module_resource('base', 'tests', 'asset_tag.xml'),
```
Comment (rdeodoo) [inline odoo/addons/base/tests/test_ir_asset.py:52]: linter

File: addons/website/tests/test_theme_ir_asset.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+
+@tagged('-at_install', 'post_install')
+class TestThemeAsset(TransactionCase):
+
+    def test_00_asset_tag(self):
+        """Verify that assets defined with the <asset> tag are properly imported"""
+        # Load new records
+        convert_file(self.cr, 'theme_default',
+            get_module_resource('website', 'tests', 'asset_tag.xml'),
```
Comment (rdeodoo) [inline addons/website/tests/test_theme_ir_asset.py:16]: linter

File: addons/website/tests/test_theme_ir_asset.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        prepend_asset.bundle = 'changed'
+        prepend_asset.path = 'theme_default/tests/changed.scss'
+        asset_with_extra_field.sequence = 3
+        active_switch_asset_reset.active = False
+        active_switch_asset_ignore.active = False
+        inactive_switch_asset.active = True
+
+        # Update records
+        convert_file(self.cr, 'theme_default',
+            get_module_resource('website', 'tests', 'asset_tag.xml'),
```
Comment (rdeodoo) [inline addons/website/tests/test_theme_ir_asset.py:51]: linter

**Review by rdeodoo — CHANGES_REQUESTED** [review]
A few linter errors (all the same kind so maybe check why this one is not catch in your IDE) + need to move tests data

> But it's almost a copy/paste of each other. If you prefer, I can put both inside website and factorize a bit.

It's fine, it's only 50 lines of code in a test which are not exactly the same (as one is testing theme.ir.asset on top of it), as you prefer

File: addons/website/tests/asset_tag.xml (hunk 3 above)
Comment (bso-odoo) [reply addons/website/tests/asset_tag.xml:4]: @rdeodoo Should I also move the existing `website/tests/template_qweb_test.xml` to `test_website` ? (maybe in a distinct PR)
Note that those files are not loaded in a normal module load - they are not part of `website`'s `data` manifest entry. They are only used during tests that explicitly trigger their import.

Comment (rdeodoo) [reply addons/website/tests/asset_tag.xml:4]: You can indeed, but in master. This one is in `website` because I introduced `test_website` [introduced](https://github.com/odoo/odoo/commit/ef03db9edd9472201cb2c08a32d20ff0f33a5fdf) after that [file](https://github.com/odoo/odoo/commit/9cd982bcc811cacb42f5c08db139043d2734b891).

> Note that those files are not loaded in a normal module load - they are not part of website's data manifest entry. They are only used during tests that explicitly trigger their import.

Still, it's noising the file hierarchy for no good reason (it's really not that big of a deal but since we have a module just for that..)

Comment (rdeodoo) [reply addons/website/tests/asset_tag.xml]: Moved to https://github.com/odoo/odoo/pull/109349

**Review by chevalierv — COMMENTED** [review]
Hello @qsm-odoo 

Why do you intend to make this change in stable, especially 17.0, which is kinda old now? Can't this break customization?

**Comment by qsm-odoo** [comment]
@AntoineVDV TBH, it's been a long time since I touched the documentation. I have always desired to tackle it seriously... but between the multiple refactorings, I guess it was never meant to be.

I opened a PR to be merged in the same batch as this one: https://github.com/odoo/documentation/pull/14984 -> would you want to have a look and approve it? Thanks.

**Review by xmo-odoo — COMMENTED** [review]
@robodoo override=ci/security

**Review by qsm-odoo — APPROVED** [review]
@robodoo r+
